### DISCUSSION TITLE
Case sensitive module names

### DISCUSSION THREAD

**Discussion by bengosu** [miurahr/aqtinstall#379] [General]
Hi @miurahr, thanks for this very helpful tool!
I want to give you some feedback to spare some of your users of the headache I've got for the last couple of days.

I was trying to install Qt with QtWebEngine  module and failed with only a generic "Some of specified modules are unknown" error. I tried looking for ways of seeing the available modules for that particular Qt version but couldn't find any. I found some documentation here: https://aqtinstall.readthedocs.io/_/downloads/en/latest/pdf/, describing just that  "aqt list-qt windows android --modules 5.15.2 # Print modules available" but it seems this documentation is for an older version as the list-qt is no longer available and using just "list" doesn't list the modules as in this example.

I finally figured out my problem, I was running the command with "QtWebEngine" as module name and apparently the module names are case sensitive and accept only lower case letters. 
I figured this out by accident as there is no mention of this anywhere.

So my feedback is this:
- if you cannot fix the caps sensitive problem, have a mention somewhere that the module names are case sensitive and are supposed to be lower caps 
- make sure you have a way for the users to see which modules are available for which version

Please see this as feedback for your great tool, I don't want to sound like an entitled brat :)

Thank you again for your great work!

**Comment by pzhlkj6612** [comment]
Hi.

The document your link points to says:

> **From version 2.0.0**, sub commands are changed. The previous versions of these sub commands have been retained for
backwards compatibility, but are no longer recommended
> ...
> | New sub commands | Legacy sub commands | Note
> |:- |:- |:-
> | ... | ... | ...
> | list-qt

So, what version of aqtinstall are you using?

**Reply by bengosu** [reply]
Hi,

I'm using it in an Azure Pipeline, using this article as a guideline: https://lostdomain.org/2019/12/27/using-azure-devops-pipelines-with-qt/
As mentioned in the article, I'm installing it with this command `python -m pip install aqtinstall` and get this as an output:

Collecting aqtinstall
  Downloading aqtinstall-1.2.5-py2.py3-none-any.whl (34 kB)
Collecting semantic-version
  Downloading semantic_version-2.8.5-py2.py3-none-any.whl (15 kB)
Collecting py7zr>=0.15.1
  Downloading py7zr-0.16.1-py3-none-any.whl (65 kB)
Collecting patch>=1.16
  Downloading patch-1.16.zip (12 kB)
Collecting texttable
  Downloading texttable-1.6.4-py2.py3-none-any.whl (10 kB)
Collecting requests
  Downloading requests-2.26.0-py2.py3-none-any.whl (62 kB)
Collecting bcj-cffi<0.6.0,>=0.5.1
  Downloading bcj_cffi-0.5.1-cp39-cp39-macosx_10_9_x86_64.whl (19 kB)
Collecting pycryptodomex>=3.6.6
  Downloading pycryptodomex-3.10.1-cp35-abi3-macosx_10_9_x86_64.whl (1.5 MB)
Collecting pyppmd>=0.14.0
  Downloading pyppmd-0.16.1-cp39-cp39-macosx_10_9_x86_64.whl (36 kB)
Collecting brotli>=1.0.9
  Downloading Brotli-1.0.9-cp39-cp39-macosx_10_9_x86_64.whl (421 kB)
Collecting multivolumefile>=0.2.3
  Downloading multivolumefile-0.2.3-py3-none-any.whl (17 kB)
Collecting pyzstd<0.15.0,>=0.14.4
  Downloading pyzstd-0.14.4-cp39-cp39-macosx_10_9_x86_64.whl (431 kB)
Collecting cffi>=1.14.0
  Downloading cffi-1.14.6-cp39-cp39-macosx_10_9_x86_64.whl (176 kB)
Collecting pycparser
  Downloading pycparser-2.20-py2.py3-none-any.whl (112 kB)
Collecting certifi>=2017.4.17
  Using cached certifi-2021.5.30-py2.py3-none-any.whl (145 kB)
Collecting idna<4,>=2.5
  Downloading idna-3.2-py3-none-any.whl (59 kB)
Collecting urllib3<1.27,>=1.21.1
  Downloading urllib3-1.26.6-py2.py3-none-any.whl (138 kB)
Collecting charset-normalizer~=2.0.0
  Downloading charset_normalizer-2.0.4-py3-none-any.whl (36 kB)
Using legacy 'setup.py install' for patch, since package 'wheel' is not installed.
Installing collected packages: pycparser, cffi, urllib3, texttable, pyzstd, pyppmd, pycryptodomex, multivolumefile, idna, charset-normalizer, certifi, brotli, bcj-cffi, semantic-version, requests, py7zr, patch, aqtinstall
    Running setup.py install for patch: started
    Running setup.py install for patch: finished with status 'done'
Successfully installed aqtinstall-1.2.5 bcj-cffi-0.5.1 brotli-1.0.9 certifi-2021.5.30 cffi-1.14.6 charset-normalizer-2.0.4 idna-3.2 multivolumefile-0.2.3 patch-1.16 py7zr-0.16.1 pycparser-2.20 pycryptodomex-3.10.1 pyppmd-0.16.1 pyzstd-0.14.4 requests-2.26.0 semantic-version-2.8.5 texttable-1.6.4 urllib3-1.26.6

So I'm assuming I'm using version 1.2.5.

**Reply by pzhlkj6612** [reply]
> - make sure you have a way for the users to see which modules are available for which version

This function has been implemented and will be in the version of 2.0.0 (not yet released).

> So I'm assuming I'm using version 1.2.5.

OK, that's true.

To use version 1.2.5, you should read https://aqtinstall.readthedocs.io/_/downloads/en/v1.2.5/pdf/ .

**Reply by bengosu** [reply]
How about case insensitive module names, can this be implemented?

**Reply by pzhlkj6612** [reply]
> How about case insensitive module names, can this be implemented?

I guess that this may not be a hard work, but I don't know if it will bring some side effects...

And unfortunately, you may have noticed that the version 2.0.0 is currently in the RC (Release Candidate) stage, so no new features will be added.

<br>

@miurahr , @ddalcino , what do you think about it?

**Reply by pzhlkj6612** [reply]
@bengosu , I guess you might want to try the unreleased version. For [2.0.0rc2](https://pypi.org/project/aqtinstall/2.0.0rc2/), you can install it by executing `pip install aqtinstall==2.0.0rc2` . The syntax for specifying the version number can be found in [Installing Packages § User Guide - pip documentation](https://pip.pypa.io/en/stable/user_guide/#installing-packages).

**Reply by pzhlkj6612** [reply]
Sorry, it's the unstable version, not unreleased.

**Comment by ddalcino** [comment]
> > How about case insensitive module names, can this be implemented?

> I guess that this may not be a hard work, but I don't know if it will bring some side effects...

Hi @bengosu , I'm sorry to hear about the trouble you were having.

Yes, case-insensitive module names can be added easily. At this point, all the module names are written completely in lowercase in the Updates.xml files in the Qt repo; you can confirm this by searching for uppercase letters in `aqt/combinations.json`. I do not foresee any side effects.

For that matter, we could add case-insensitivity to the architectures, tool names, and tool variants as well. If we add case-insensitivity to any of these, it would be inconsistent not to add case-insensitivity to all the others.

However, I think the real problem here is that @bengosu was not able to access `aqt list-qt`. You cannot expect to use `aqt install-*` successfully without using the `list-*` subcommands. If you're trying to install some module, architecture, or tool that is not available for your host/target/version combination, the command is going to fail, case-sensitivity or not. There's no way to know the right command to use without consulting `aqt list-*` or reading the 'Updates.xml' files at https://download.qt.io/online/qtsdkrepository.

I'm not trying to blame @bengosu for this problem; the easiest documentation to find is for version 2.0, and the easiest version of aqt to install is 1.2. @bengosu is not the first person to have this issue, and I don't see an easy solution to it.

Maybe I'm wrong about this, but I don't think it makes a lot of sense to add case-insensitivity. If you disagree, I can file a PR to add it, but the changes will most likely go into some version of 2.0, not 1.2.

**Reply by bengosu** [reply]
Hi @ddalcino, thank you for the detailed response!

Finding the "Updates.xml" and copying the "qtwebebgine" module name from it, to make sure I don't misspell it, was how I figured out what my problem was. When writing the `aqt install` command I also copied the "QtWebEngine" module name from my local installation, again, to make sure I don't misspell it :D.

If the case-insensitive fix is not something you can easily do, you could at least add something about this in the README.
Something like "Please note that the module names should all be lower case".

Thank you again for taking the time to respond.

**Reply by ddalcino** [reply]
We can add a note in the documentation, if you think that would be helpful. The README is undergoing a big redesign right now as we get ready for the 2.0 release; you can follow it here: #377. We're trying to keep the README brief and to the point, so we don't have two parallel sets of instructions that we need to update. As it stands, it looks like all discussion of modules is being removed entirely from the README.

**Reply by ddalcino** [reply]
> Finding the "Updates.xml" and copying the "qtwebebgine" module name from it, to make sure I don't misspell it, was how I figured out what my problem was. When writing the `aqt install` command I also copied the "QtWebEngine" module name from my local installation, again, to make sure I don't misspell it :D.

I'm sorry you had to do that; we don't want our users to have to go through that!

> If the case-insensitive fix is not something you can easily do, you could at least add something about this in the README.
> Something like "Please note that the module names should all be lower case".

I added a suggestion here: https://github.com/miurahr/aqtinstall/pull/377#discussion_r698733733; it will only go in if there's room.

> Thank you again for taking the time to respond.

No problem!

**Reply by pzhlkj6612** [reply]
@bengosu ,

> Finding the "Updates.xml" and copying the "qtwebebgine" module name from it, to make sure I don't misspell it, was how I figured out what my problem was...

Could you please post the URL to this "Updates.xml" file? I think we could add a tip to README about this misspelled module name.

**Reply by bengosu** [reply]
I thing it might have been this one https://download.qt.io/online/qtsdkrepository/mac_x64/desktop/qt5_51211/Updates.xml

**Reply by pzhlkj6612** [reply]
> I thing it might have been this one https://download.qt.io/online/qtsdkrepository/mac_x64/desktop/qt5_51211/Updates.xml

Hmm, after checking every Updates.xml file from `https://download.qt.io/online/qtsdkrepository/(windows_x86|mac_x64|linux_x64)/desktop/qt[56]_(\d+)/Updates.xml` , I still cannot find the  misspelled module name "qtwebebgine".

**Reply by bengosu** [reply]
Nothing is misspelled, the problem is the fact that the `aqt install` command is case sensitive. 

I used QtWebEngine as module name and I got module not found message so I copied the module name from the Updates.xml file to make sure I didn't misspell it and it worked. In Updates.xml the module name is qtwebengine so that's how I figured out that the `aqt install` command is case sensitive,

**Reply by pzhlkj6612** [reply]
Oh, I‘ve misunderstood your [previous comment](https://github.com/miurahr/aqtinstall/discussions/379#discussioncomment-1255002) because the module name was misspelled by you.

```diff
- qtwebebgine
+ qtwebengine
        ^
```

It's alright now, and sorry for the noise.

**Reply by bengosu** [reply]
Oh, sorry about that! :D

**Comment by miurahr** [comment]
> Yes, case-insensitive module names can be added easily. At this point, all the module names are written completely in lowercase in the `Updates.xml` files in the Qt repo; you can confirm this by searching for uppercase letters in `aqt/combinations.json`. I do not foresee any side effects.
> For that matter, we could add case-insensitivity to the architectures, tool names, and tool variants as well. If we add case-insensitivity to any of these, it would be inconsistent not to add case-insensitivity to all the others.

Every parameters for architectures, tool names, and module names are defined by Qt team, but not by us. There is no guarantee for its lower case rule. So we observe current repository and provide lists to users through `list-qt` sub command. 

According to a principle of *no surprise* for users, it is natural that `install-qt` sub command accepts  keywords `list-qt` shows.
And it may be better that `install-qt` suggest a lower case name when it is found in `aqt/combinations.json`.

In the case of here, when user put `-m QtWebEngine` and it is not found in `aqt/combinations.json` with case sensitivity, but found in lower case, we can put warning `do you mean qtwebengine?`
